### ISSUE TITLE
Pypi release for pytorch 1.7

### ISSUE THREAD

**Issue by lgvaz** [rwightman/efficientdet-pytorch#114]
Hi Ross, me again :sweat_smile: 

Any plans for a pypi release with support for pytorch 1.7?

**Comment by rwightman** [comment]
@lgvaz  hi hi, I was originally planning on waiting a bit, see if any issues came up given how many changes were made... not clear if/how well it's been working for others. Have you tried the master? Switched to using the head reset mechanism I added there (proper re-init)?

I personaly don't think 1.7 is a good reason given that the CUDA 11 release has a problematic cuDNN version. I'd wait for 1.7.1 .. but I guess fastai already made the switch.

If you confirm you've had no issues with the current master I can speed it along and get the release up tomorrow/Saturday.

**Comment by lgvaz** [comment]
I tried master and my tests are failing with some shapes mismatch on the loss function that I was not able to figure out. I was not sure if I still had to adapt my code to the some breaking changes on effdet or if it was an internal bug. 

[Here](https://github.com/airctic/icevision/issues/533) is the issue describing the problems I'm facing, any ideas?

My main interest in updating to torch 1.7 is because colab made the switch and when installing `pip install icevision[all]` torch 1.6 is being installed without GPU support :cry: (so currently we're having to run the additional command `pip install torch==1.6.0+cu101 torchvision==0.7.0+cu101 -f https://download.pytorch.org/whl/torch_stable.html` there)

**Comment by rwightman** [comment]
@lgvaz in the error it looks like the anchor targets are using coco defaults but your inputs are using different values. You might want to check your model creation and make sure you're setting the image_size and num_classes in the model's copy of the config before wrapping in a Bench. I changed some of the config handling in the model/factory. None of that is PyTorch 1.7 relaed.

Only PyTorch 1.7 specific issue is an old style string format that got picked up as a remainder op. I believe it's a torchscript bug. Fixed by changing to an f-string (on master).

**Comment by lgvaz** [comment]
> I changed some of the config handling in the model/factory  

That was exactly it, thanks! I was still passing config to `DetBenchTrain`

Alright, training seems to be working now correctly now, you can check [here](https://app.reviewnb.com/airctic/icevision/pull/532/files/) it you want (it's the quickstart notebook)

---
Weights trained with the old version of the library are also correctly loading, there is just one small difference, before I was getting two predictions for a specific image (I'm manually filtering `scores > .5`):

```python
[{'scores': array([0.99850035, 0.8992443 ], dtype=float32), 'labels': array([2, 3]), 'bboxes': [<BBox (xmin:88.84365844726562, ymin:77.79306030273438, xmax:224.62063598632812, ymax:341.6015930175781)>, <BBox (xmin:149.5955810546875, ymin:289.87567138671875, xmax:460.33990478515625, ymax:381.52923583984375)>]}]
```
![image](https://user-images.githubusercontent.com/16520544/98380450-5c57bf80-2027-11eb-911f-45f5d5718a98.png)

Now I'm getting only one result, and the score is a bit different:

```python
[{'scores': array([0.9598645], dtype=float32), 'labels': array([2]), 'bboxes': [<BBox (xmin:88.8436508178711, ymin:77.79306030273438, xmax:224.62060546875, ymax:341.6015930175781)>]}]
```

![image](https://user-images.githubusercontent.com/16520544/98380239-100c7f80-2027-11eb-8d38-debcf822f4dc.png)

Maybe this is also related to some change in the internal config handling? It's worth saying that to create `DetBenchPredict` I'm just unwrapping from `DetBenchTrain`:
```python
bench = DetBenchPredict(unwrap_bench(model))
```

**Comment by rwightman** [comment]
@lgvaz did the missing box completely vanish from the output or was it just filtered because it fell below < 0.5? 

Did you compare my master with both PyTorch 1.6 and 1.7 to see if the difference is PT 1.7? I general I've noticed more numeric differences in PT 1.7 than any recent version. My scores in other models, classification, etc all had a bigger change than usual. I haven't done extensive testing on 1.7 w/ EffDet yet, just verified validation on one model and early coco training convergence looked alright.

**Comment by rwightman** [comment]
Another thing, probably worth looking at the updated model creation factory ...https://github.com/rwightman/efficientdet-pytorch/blob/master/effdet/factory.py#L34-L49 

For changing the head, I'd use the reset function I added, it makes sure the head weights are re-initialized properly. Code that was just replacing the head without call re-init was leaving the heads init to non-optimial values in the past. It trained, but would have converged a bit slower.

I don't see any issues creating the predict from unwraped train model. But I'm not seeing the need to create both a DetBenchPredict and Train. Unless I missed a use case.... Predict is for just predicting, train takes into consideration the need for predictions while validating... when model is in .eval() mode it outputs loss + predictions, in train, just the loss. Predict model never creates the loss fn or outputs the loss.

**Comment by lgvaz** [comment]
> For changing the head, I'd use the reset function I added

Thanks for the tip! I'm going to add that, actually I'll switch to use your model creation factory :smile: 

> But I'm not seeing the need to create both a DetBenchPredict and Train

I'm doing something sub-optimal :sweat_smile:, I wanted to hide the creation of bench from my users, so I made `DetBenchTrain` the default. When the user wants to predict I then wrap the model with the `DetBenchPredict`. Does this creates too much computational overhead? Another option would be to assume that the standard is the unwrapped model and wrap correspondingly while training/inference.

---
I investigated the numerical differences deeper and found it's being cause by a bug in my code, I traced the error back to a specific line change but I'm not sure why it's happening.

The error is caused by how I create the model, as I said before I'll start using your model factory but I'll post the error here for completeness. This is how I'm currenly creating the model:

```python
    config = get_efficientdet_config(model_name=model_name)

    net = EfficientDet(config=config, pretrained_backbone=False)
    if pretrained:
        if not config.url:
            raise RuntimeError(f"No pretrained weights for {model_name}")
        state_dict = torch.hub.load_state_dict_from_url(
            config.url, map_location=torch.device("cpu")
        )
        net.load_state_dict(state_dict)

--> config = net.config  # config gets deepcopied inside Efficientdet
    config.num_classes = num_classes
    config.image_size = (img_size, img_size) if isinstance(img_size, int) else img_size
    net.class_net = HeadNet(
        config=config,
        num_outputs=num_classes,
    )

   model_bench = DetBenchTrain(model=net)
``` 

The line marked with the arrow is the key, if I leave it uncommented I get the following results:
```python
array([0.9598645 , 0.22345279, 0.05992804, 0.0233851...
```
But if I comment it out:
```python
array([0.9598645 , 0.9174616 , 0.68157065, 0.5030457
```

Big big difference!!! But the first result stays the same! I'm guessing config gets modified internally and when the line is commented out the changes don't get reflected in the config stored in `net`? Which approach is the correct one?

**NOTE:** I'll transition to the model factory and post a new comment with the results :grinning:

**Comment by lgvaz** [comment]
Alright Ross, this is amazing, before the head reset mechanism I had to train only the head of the model for some epochs just to get to an acceptable loss:

![image](https://user-images.githubusercontent.com/16520544/98429675-e125f600-2086-11eb-8b20-585a725a6497.png)

Now it just starts with a loss that is even lower than that of the previous final epoch!

![image](https://user-images.githubusercontent.com/16520544/98429703-116d9480-2087-11eb-9523-62af0d487b44.png)

**AMAZING** update!!!!

**Comment by lgvaz** [comment]
The only downside is that creating the model using the model factory does not yield compatible weights with the way I was previously creating models, so that is a breaking change for me (still well worth it)

**Comment by rwightman** [comment]
@lgvaz weights not being compatible is odd, means some aspect of the model config changed (that shouldn't have)...

I'm currently tweaking the config handling a bit to make it less prone to mistakes.

**Comment by lgvaz** [comment]
> weights not being compatible is odd

Probably a mistake on my part right? On how I was creating the model versus how the model factory creates

**Comment by lgvaz** [comment]
Now that these issues are addressed, can you make a pypi release? :sweat_smile: 

If you are still concerned with pytorch 1.7 you can create a pre-release so users will still get `0.1.6` when normally installing from pypi.

**Comment by rwightman** [comment]
@lgvaz I just pushed a commit that sets the config to read only and limits the use of config in forward() call chain so that it doesn't break torchscript, I also finally figured out how to get torchscript to work for the model and suprsingly the bench. I did this to try and enforce or at least indicate that the config is intended not to be modified after it's passed to the model unless you REALLY know what you're doing. It certainly cannot be changed after the model is wrapped in bench, that'll break everything.

It'd be great if you can run what I just pushed and verify no additional breaks... I've tested some training and validation of a number of models. I'll push the pypi release tonight if you confirm.

**Comment by lgvaz** [comment]
I tested training with `tf_efficientdet_lite0` on the fridge and raccoon dataset with fastai and lightning and can confirm everything is in order for me :smiley:

**Comment by rwightman** [comment]
@lgvaz k, just pushed 0.2.1 ... let me know if any issues come up

**Comment by lgvaz** [comment]
Thanks Ross!! I'll update tomorrow and close the issue if everything works well =)

**Comment by lgvaz** [comment]
Sorry! I forgot to close
